R/position-sina.r

```diff
@@ -0,0 +1,221 @@
+#' Point density based jitter.
+#'
+#' @family position adjustments
+#' @param binwidth Binning factor. The range of the values in \code{aes(y)}
+#' are binned in windows of length \code{(max(y) - min(y)) * y_fraction}.
+#' Samples within the same bin belong to the same "neighbourhood".
+#' @param scale logical. When set to \code{TRUE} x-coordinate widths across all
+#' groups are scaled based on the densiest are in the plot. Default: \code{TRUE}
+#' @param neighbour_limit if the samples within the same y-axis bin are more
+#' than neighbour_limit, the samples's X coordinates will be adjusted.
+#' @param method choose the method to spread the samples within the same
+#' neighbourhood along the x-axis. Available methods: "density",
+#' "neighbourhood" (can be abbreciated, e.g. "d"). See \code{Details}.
+#' @param adjust adjusts the bandwidth of the density kernel when
+#' \code{method == "density"} (see \code{\link[stats]{density}}) or the spread
+#' of the samples within the same group along the x-axis.
+#' @export
+#' @examples
+#' # See geom_sina for examples.
+position_sina <- function(binwidth = 0.2,
+                          scale = TRUE,
+                          neighbour_limit = 1,
+                          method = c("density", "neighbourhood"),
+                          adjust = 1) {
+  ggproto(NULL, PositionSina,
+          binwidth = binwidth,
+          scale = scale,
+          neighbour_limit = neighbour_limit,
+          method = match.arg(method),
+          adjust = adjust
+  )
+}
+
+#' @rdname ggplot2-ggproto
+#' @format NULL
+#' @usage NULL
+#' @export
+PositionSina <- ggproto("PositionSina", Position,
+
+  required_aes = c("x", "y"),
+
+  setup_params = function(self, data) {
+
+    if (is.double(data$x) && any(data$x != data$x[1L])) {
+      stop(
+        "Continuous x aesthetic -- maybe try aes(factor(x))?",
+        call. = FALSE)
+    }
+
+    list(
+      binwidth = self$binwidth,
+      scale = self$scale,
+      neighbour_limit = self$neighbour_limit,
+      method = self$method,
+      adjust = self$adjust
+    )
+  },
+
+  setup_data = function(self, data, params) {
+    data = remove_missing(data, FALSE,
+      c("x", "y", "ymin", "ymax", "xmin", "xmax"))
+    data
+  },
+
+  compute_layer = function(data, params, panel) {
+    trans_x <- function(x) .sina(x, data, binwidth = params$binwidth,
+                                 scale = params$scale,
+                                 neighbour_limit = params$neighbour_limit,
+                                 method = params$method,
+                                 adjust = params$adjust)
+    trans_y <- NULL
+
+    transform_position(data, trans_x, trans_y)
+  }
+)
+
+.sina <- function(x, data, binwidth, scale, neighbour_limit, method, adjust) {
+
+
+  ### Initialise variables
+  #x-axis transpose vector
+  trans_x <- list()
+
+  #assing groups as factors of x
+  groups <- factor(x)
```

Comment:
The groups are available at `data$group` and it would guarantee that you don't run into edgecase bugs because a user became very creative in how they used `aes()`.